### PR TITLE
Make more APIs follow the SDL_GetStringRule.

### LINKED ISSUE #10229: More functions that could use the SDL_GetStringRule...
There are a handful of things that allocate strings for return values that we expect the app to free.

Should these be changed to return `const char *` and follow the SDL_GetStringRule, so the app doesn't have to free it, they just have to use it and be done?  (or copy it if they want to keep it around).

These are the things that return `char *` with obvious non-allocations (SDL_getenv, SDL_strlwr, etc) filtered out.

```
SDL_stdinc.h:2895:extern SDL_DECLSPEC char *SDLCALL SDL_iconv_string(const char *tocode,
SDL_clipboard.h:75:extern SDL_DECLSPEC char * SDLCALL SDL_GetClipboardText(void);
SDL_clipboard.h:120:extern SDL_DECLSPEC char * SDLCALL SDL_GetPrimarySelectionText(void);
SDL_filesystem.h:83:extern SDL_DECLSPEC char *SDLCALL SDL_GetBasePath(void);
SDL_filesystem.h:140:extern SDL_DECLSPEC char *SDLCALL SDL_GetPrefPath(const char *org, const char *app);
SDL_filesystem.h:243:extern SDL_DECLSPEC char *SDLCALL SDL_GetUserFolder(SDL_Folder folder);
SDL_filesystem.h:380:extern SDL_DECLSPEC char **SDLCALL SDL_GlobDirectory(const char *path, const char *pattern, SDL_GlobFlags flags, int *count);
SDL_storage.h:404:extern SDL_DECLSPEC char **SDLCALL SDL_GlobStorageDirectory(SDL_Storage *storage, const char *path, const char *pattern, SDL_GlobFlags flags, int *count);
SDL_gamepad.h:402:extern SDL_DECLSPEC char ** SDLCALL SDL_GetGamepadMappings(int *count);
SDL_gamepad.h:418:extern SDL_DECLSPEC char * SDLCALL SDL_GetGamepadMappingForGUID(SDL_JoystickGUID guid);
SDL_gamepad.h:438:extern SDL_DECLSPEC char * SDLCALL SDL_GetGamepadMapping(SDL_Gamepad *gamepad);
SDL_gamepad.h:663:extern SDL_DECLSPEC char *SDLCALL SDL_GetGamepadInstanceMapping(SDL_JoystickID instance_id);
```

### PR DESCRIPTION
**Pull request by icculus**
This is the work to make more things that return allocated strings follow the SDL_GetStringRule. This is still a draft, because I'm going to change these functions one commit at a time and have several to go right now.

First one up was SDL_GetBasePath(), which resulted in more improvements than just memory management rules.

Fixes #10229.

### CODE REVIEW

File: docs/README-migration.md [hunk 1]
```diff
@@ -802,10 +802,9 @@ On Haiku OS, SDL no longer sets the current working directory to the executable'
 
 ```c
 {
-    char *path = SDL_GetBasePath();
+    const char *path = SDL_GetBasePath();
     if (path) {
-        chdir(path);
-        SDL_free(path);
+        chdir(path);  /* don't free this string, SDL owns it. */
```
Comment (slouken) [inline docs/README-migration.md:807]: I don't think you need the comment. We're not highlighting the string rule here, we're just showing how to get the SDL2 behavior.

File: include/SDL3/SDL_filesystem.h [hunk 2]
```diff
@@ -68,8 +68,8 @@ extern "C" {
  * The returned path is guaranteed to end with a path separator ('\\' on
  * Windows, '/' on most other platforms).
  *
- * The pointer returned is owned by the caller. Please call SDL_free() on the
- * pointer when done with it.
+ * The returned string follows the SDL_GetStringRule. This is different than
+ * SDL2! Do _not_ call SDL_free on the returned pointer!
```
Comment (slouken) [inline include/SDL3/SDL_filesystem.h:72]: I don't think you need this warning here, we're returning const char *, which a compiler will warn about and won't let you free.

File: docs/README-migration.md (hunk 1 above)
Comment (icculus) [reply docs/README-migration.md]: Fixed in bbffc3a4940cddefcc85a4fcd6312ef8011bd296.

File: include/SDL3/SDL_filesystem.h (hunk 2 above)
Comment (icculus) [reply include/SDL3/SDL_filesystem.h]: Fixed in bbffc3a4940cddefcc85a4fcd6312ef8011bd296.

**Comment by slouken** [comment]
Let's prioritize this? I'd like to get it in for the next Steam update.

Thanks!

**Comment by icculus** [comment]
Okay, I think this is all of them. I've rebased against the latest in revision control to clean up a conflict, so if the builders like it, it can be merged.

sdl2-compat changes:

```diff
diff --git a/src/sdl2_compat.c b/src/sdl2_compat.c
index 800f359..d73e17e 100644
--- a/src/sdl2_compat.c
+++ b/src/sdl2_compat.c
@@ -147,6 +147,12 @@ extern "C" {
     SDL_DECLSPEC rc SDLCALL SDL_##oldfn params { ret SDL3_##newfn args; }
 #include "sdl3_syms.h"
 
+/* Things that are the same in SDL3, except they now follow the SDL_GetStringRule (SDL3 owns the returned string, caller does not, so we make a copy here). */
+/* Note that these need to be in sdl3_syms.h with the _non-const_ return type to make this work! */
+#define SDL3_SYM_GETSTRINGRULE(rc,fn,params,args,ret) \
+    SDL_DECLSPEC rc SDLCALL SDL_##fn params { const rc retval = SDL3_##fn args; ret retval ? SDL3_strdup(retval) : NULL; }
+#include "sdl3_syms.h"
+
 
 /* these are macros (etc) in the SDL headers, so make our own. */
 #define SDL3_AtomicIncRef(a)  SDL3_AtomicAdd(a, 1)
@@ -8181,6 +8187,23 @@ SDL_GameControllerMappingForIndex(int idx)
     return retval;
 }
 
+/* Was renamed _and_ follows the SDL_GetStringRule, so we had to fill this in manually. */
+SDL_DECLSPEC char* SDLCALL
+SDL_GameControllerMappingForGUID(SDL_JoystickGUID guid)
+{
+    const char *retval = SDL3_GetGamepadMappingForGUID(guid);
+    return retval ? SDL3_strdup(retval) : NULL;
+}
+
+/* Was renamed _and_ follows the SDL_GetStringRule, so we had to fill this in manually. */
+SDL_DECLSPEC char* SDLCALL
+SDL_GameControllerMapping(SDL_GameController *controller)
+{
+    const char *retval = SDL3_GetGamepadMapping(controller);
+    return retval ? SDL3_strdup(retval) : NULL;
+}
+
+
 SDL_DECLSPEC SDL_GameController* SDLCALL
 SDL_GameControllerOpen(int idx)
 {
@@ -9096,7 +9119,6 @@ SDL_hid_enumerate(unsigned short vendor_id, unsigned short product_id)
     return retval;
 }
 
-
 static SDL_Thread *SDL2_CreateThreadWithStackSize(SDL_ThreadFunction fn, const char *name, const size_t stacksize, void *userdata, SDL_FunctionPointer pfnBeginThread, SDL_FunctionPointer pfnEndThread)
 {
     const SDL_PropertiesID props = SDL3_CreateProperties();
diff --git a/src/sdl3_syms.h b/src/sdl3_syms.h
index 68ee8c2..e01bd52 100644
--- a/src/sdl3_syms.h
+++ b/src/sdl3_syms.h
@@ -33,6 +33,9 @@
 #ifndef SDL3_SYM_VARARGS
 #define SDL3_SYM_VARARGS(rc,fn,params) SDL3_SYM(rc,fn,params,unused,unused)
 #endif
+#ifndef SDL3_SYM_GETSTRINGRULE
+#define SDL3_SYM_GETSTRINGRULE(rc,fn,params,args,ret) SDL3_SYM(rc,fn,params,args,ret)
+#endif
 
 SDL3_SYM_VARARGS(int,SetError,(SDL_PRINTF_FORMAT_STRING const char *a, ...))
 SDL3_SYM_VARARGS(void,Log,(SDL_PRINTF_FORMAT_STRING const char *a, ...))
@@ -115,7 +118,7 @@ SDL3_SYM_RENAMED(SDL_bool,AtomicCASPtr,AtomicCompareAndSwapPointer,(void **a, vo
 SDL3_SYM_PASSTHROUGH(void*,AtomicSetPtr,(void **a, void *b),(a,b),return)
 SDL3_SYM_PASSTHROUGH(void*,AtomicGetPtr,(void **a),(a),return)
 SDL3_SYM_PASSTHROUGH(int,SetClipboardText,(const char *a),(a),return)
-SDL3_SYM_PASSTHROUGH(char*,GetClipboardText,(void),(),return)
+SDL3_SYM_GETSTRINGRULE(char*,GetClipboardText,(void),(),return)
 SDL3_SYM_PASSTHROUGH(SDL_bool,HasClipboardText,(void),(),return)
 SDL3_SYM_PASSTHROUGH(int,GetCPUCount,(void),(),return)
 SDL3_SYM_PASSTHROUGH(int,GetCPUCacheLineSize,(void),(),return)
@@ -150,11 +153,11 @@ SDL3_SYM(void,SetEventFilter,(SDL_EventFilter a, void *b),(a,b),)
 SDL3_SYM(void,FilterEvents,(SDL_EventFilter a, void *b),(a,b),)
 SDL3_SYM(void *,AllocateEventMemory,(size_t a),(a),return)
 SDL3_SYM(Uint32,RegisterEvents,(int a),(a),return)
-SDL3_SYM_PASSTHROUGH(char*,GetBasePath,(void),(),return)
-SDL3_SYM_PASSTHROUGH(char*,GetPrefPath,(const char *a, const char *b),(a,b),return)
+SDL3_SYM_GETSTRINGRULE(char*,GetBasePath,(void),(),return)
+SDL3_SYM_GETSTRINGRULE(char*,GetPrefPath,(const char *a, const char *b),(a,b),return)
 SDL3_SYM_RENAMED(int,GameControllerAddMapping,AddGamepadMapping,(const char *a),(a),return)
-SDL3_SYM_RENAMED(char*,GameControllerMappingForGUID,GetGamepadMappingForGUID,(SDL_JoystickGUID a),(a),return)
-SDL3_SYM_RENAMED(char*,GameControllerMapping,GetGamepadMapping,(SDL_GameController *a),(a),return)
+SDL3_SYM(const char*,GetGamepadMappingForGUID,(SDL_JoystickGUID a),(a),return)
+SDL3_SYM(const char*,GetGamepadMapping,(SDL_GameController *a),(a),return)
 SDL3_SYM_RENAMED(const char*,GameControllerName,GetGamepadName,(SDL_GameController *a),(a),return)
 SDL3_SYM_RENAMED(SDL_bool,GameControllerGetAttached,GamepadConnected,(SDL_GameController *a),(a),return)
 SDL3_SYM_RENAMED(SDL_Joystick*,GameControllerGetJoystick,GetGamepadJoystick,(SDL_GameController *a),(a),return)
@@ -853,3 +856,4 @@ SDL3_SYM(SDL_Palette *,CreateSurfacePalette,(SDL_Surface *a),(a),return)
 #undef SDL3_SYM_PASSTHROUGH
 #undef SDL3_SYM_RENAMED
 #undef SDL3_SYM_VARARGS
+#undef SDL3_SYM_GETSTRINGRULE
```

File: docs/README-macos.md [hunk 3]
```diff
@@ -228,10 +228,10 @@ Some things that may be of interest about how it all works...
...
 
 In SDL 1.2, the working directory of your SDL app is by default set to its
-parent, but this is no longer the case in SDL 2.0. SDL2 does change the
-working directory, which means it'll be whatever the command line prompt
-that launched the program was using, or if launched by double-clicking in
-the finger, it will be "/", the _root of the filesystem_. Plan accordingly!
+parent, but this is no longer the case in SDL 2.0 and later. SDL2 does not
+change the working directory, which means it'll be whatever the command line
+prompt that launched the program was using, or if launched by double-clicking
+in the Finder, it will be "/", the _root of the filesystem_. Plan accordingly!
```
Comment (slouken) [inline docs/README-macos.md:234]: Did you mean to have this as part of this PR?

Comment (icculus) [reply docs/README-macos.md:234]: Oh, whoops, that was probably a rebase issue or something I slurped in while jumping between PRs. Will fix.

Comment (icculus) [reply docs/README-macos.md]: This is rebased out now.

**Comment by icculus** [comment]
Merge this now?

**Comment by slouken** [comment]
> Merge this now?

Yes please!

**Comment by icculus** [comment]
Okay, when my power comes back on, I'll take care of sdl2-compat (both for this and the TLS changes)

**Comment by AntTheAlchemist** [comment]
`SDL_GetJoystickGUIDString` is a bit of an oddity. Can it be changed to allocate it's own string and follow the SDL_GetStringRule?

**Comment by slouken** [comment]
Yep!

**Comment by RenaKunisaki** [comment]
Am I losing my mind, or has someone tried to scrub the existence of SDL_GetJoystickGUIDString from history? The only search results on here are a couple issues mentioning it. Even Google leads to nonexistent wiki pages and old copies of its documentation. It feels like I've uncovered the most pointless conspiracy theory ever :zany_face: 

What was it replaced with/renamed to, so I can update my app to a newer SDL version?

**Comment by AntTheAlchemist** [comment]
@RenaKunisaki in the latest source, it's been renamed to `SDL_GetJoystickGUID`

**Comment by slouken** [comment]
> Am I losing my mind, or has someone tried to scrub the existence of SDL_GetJoystickGUIDString from history? The only search results on here are a couple issues mentioning it. Even Google leads to nonexistent wiki pages and old copies of its documentation. It feels like I've uncovered the most pointless conspiracy theory ever 🤪
> 
> What was it replaced with/renamed to, so I can update my app to a newer SDL version?

I think you want SDL_GUIDToString()

**Pull request merged**